File: tests/algorithms/x/Clojure/data_structures/linked_list/singly_linked_list.clj
(ns main (:refer-clojure :exclude [empty_list length is_empty to_string insert_nth insert_head insert_tail delete_nth delete_head delete_tail get_item set_item reverse_list main]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare empty_list length is_empty to_string insert_nth insert_head insert_tail delete_nth delete_head delete_tail get_item set_item reverse_list main)

(def ^:dynamic delete_nth_i nil)

(def ^:dynamic delete_nth_res nil)

(def ^:dynamic delete_nth_val nil)

(def ^:dynamic insert_nth_i nil)

(def ^:dynamic insert_nth_res nil)

(def ^:dynamic main_del nil)

(def ^:dynamic main_i nil)

(def ^:dynamic main_lst nil)

(def ^:dynamic reverse_list_i nil)

(def ^:dynamic reverse_list_res nil)

(def ^:dynamic set_item_i nil)

(def ^:dynamic set_item_res nil)

(def ^:dynamic to_string_i nil)

(def ^:dynamic to_string_s nil)

(defn empty_list []
  (try (throw (ex-info "return" {:v {:data []}})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn length [length_list]
  (try (throw (ex-info "return" {:v (count (:data length_list))})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn is_empty [is_empty_list]
  (try (throw (ex-info "return" {:v (= (count (:data is_empty_list)) 0)})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn to_string [to_string_list]
  (binding [to_string_i nil to_string_s nil] (try (do (when (= (count (:data to_string_list)) 0) (throw (ex-info "return" {:v ""}))) (set! to_string_s (str (get (:data to_string_list) 0))) (set! to_string_i 1) (while (< to_string_i (count (:data to_string_list))) (do (set! to_string_s (str (str to_string_s " -> ") (str (get (:data to_string_list) to_string_i)))) (set! to_string_i (+ to_string_i 1)))) (throw (ex-info "return" {:v to_string_s}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn insert_nth [insert_nth_list insert_nth_index insert_nth_value]
  (binding [insert_nth_i nil insert_nth_res nil] (try (do (when (or (< insert_nth_index 0) (> insert_nth_index (count (:data insert_nth_list)))) (throw (Exception. "index out of range"))) (set! insert_nth_res []) (set! insert_nth_i 0) (while (< insert_nth_i insert_nth_index) (do (set! insert_nth_res (conj insert_nth_res (get (:data insert_nth_list) insert_nth_i))) (set! insert_nth_i (+ insert_nth_i 1)))) (set! insert_nth_res (conj insert_nth_res insert_nth_value)) (while (< insert_nth_i (count (:data insert_nth_list))) (do (set! insert_nth_res (conj insert_nth_res (get (:data insert_nth_list) insert_nth_i))) (set! insert_nth_i (+ insert_nth_i 1)))) (throw (ex-info "return" {:v {:data insert_nth_res}}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn insert_head [insert_head_list insert_head_value]
  (try (throw (ex-info "return" {:v (insert_nth insert_head_list 0 insert_head_value)})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn insert_tail [insert_tail_list insert_tail_value]
  (try (throw (ex-info "return" {:v (insert_nth insert_tail_list (count (:data insert_tail_list)) insert_tail_value)})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn delete_nth [delete_nth_list delete_nth_index]
  (binding [delete_nth_i nil delete_nth_res nil delete_nth_val nil] (try (do (when (or (< delete_nth_index 0) (>= delete_nth_index (count (:data delete_nth_list)))) (throw (Exception. "index out of range"))) (set! delete_nth_res []) (set! delete_nth_val 0) (set! delete_nth_i 0) (while (< delete_nth_i (count (:data delete_nth_list))) (do (if (= delete_nth_i delete_nth_index) (set! delete_nth_val (get (:data delete_nth_list) delete_nth_i)) (set! delete_nth_res (conj delete_nth_res (get (:data delete_nth_list) delete_nth_i)))) (set! delete_nth_i (+ delete_nth_i 1)))) (throw (ex-info "return" {:v {:list {:data delete_nth_res} :value delete_nth_val}}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn delete_head [delete_head_list]
  (try (throw (ex-info "return" {:v (delete_nth delete_head_list 0)})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn delete_tail [delete_tail_list]
  (try (throw (ex-info "return" {:v (delete_nth delete_tail_list (- (count (:data delete_tail_list)) 1))})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn get_item [get_item_list get_item_index]
  (try (do (when (or (< get_item_index 0) (>= get_item_index (count (:data get_item_list)))) (throw (Exception. "index out of range"))) (throw (ex-info "return" {:v (get (:data get_item_list) get_item_index)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn set_item [set_item_list set_item_index set_item_value]
  (binding [set_item_i nil set_item_res nil] (try (do (when (or (< set_item_index 0) (>= set_item_index (count (:data set_item_list)))) (throw (Exception. "index out of range"))) (set! set_item_res []) (set! set_item_i 0) (while (< set_item_i (count (:data set_item_list))) (do (if (= set_item_i set_item_index) (set! set_item_res (conj set_item_res set_item_value)) (set! set_item_res (conj set_item_res (get (:data set_item_list) set_item_i)))) (set! set_item_i (+ set_item_i 1)))) (throw (ex-info "return" {:v {:data set_item_res}}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn reverse_list [reverse_list_list]
  (binding [reverse_list_i nil reverse_list_res nil] (try (do (set! reverse_list_res []) (set! reverse_list_i (- (count (:data reverse_list_list)) 1)) (while (>= reverse_list_i 0) (do (set! reverse_list_res (conj reverse_list_res (get (:data reverse_list_list) reverse_list_i))) (set! reverse_list_i (- reverse_list_i 1)))) (throw (ex-info "return" {:v {:data reverse_list_res}}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn main []
  (binding [main_del nil main_i nil main_lst nil] (do (set! main_lst (empty_list)) (set! main_i 1) (while (<= main_i 5) (do (set! main_lst (insert_tail main_lst main_i)) (set! main_i (+ main_i 1)))) (println (to_string main_lst)) (set! main_lst (insert_head main_lst 0)) (println (to_string main_lst)) (set! main_del (delete_head main_lst)) (set! main_lst (:list main_del)) (println (str (:value main_del))) (set! main_del (delete_tail main_lst)) (set! main_lst (:list main_del)) (println (str (:value main_del))) (set! main_del (delete_nth main_lst 2)) (set! main_lst (:list main_del)) (println (str (:value main_del))) (set! main_lst (set_item main_lst 1 99)) (println (str (get_item main_lst 1))) (set! main_lst (reverse_list main_lst)) (println (to_string main_lst)))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (main)
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/algorithms/x/Java/data_structures/linked_list/singly_linked_list.java
public class Main {
    static class SinglyLinkedList {
        int[] data;
        SinglyLinkedList(int[] data) {
            this.data = data;
        }
        SinglyLinkedList() {}
        @Override public String toString() {
            return String.format("{'data': %s}", String.valueOf(data));
        }
    }

    static class DeleteResult {
        SinglyLinkedList list;
        int value;
        DeleteResult(SinglyLinkedList list, int value) {
            this.list = list;
            this.value = value;
        }
        DeleteResult() {}
        @Override public String toString() {
            return String.format("{'list': %s, 'value': %s}", String.valueOf(list), String.valueOf(value));
        }
    }


    static SinglyLinkedList empty_list() {
        return new SinglyLinkedList(new int[]{});
    }

    static int length(SinglyLinkedList list) {
        return list.data.length;
    }

    static boolean is_empty(SinglyLinkedList list) {
        return list.data.length == 0;
    }

    static String to_string(SinglyLinkedList list) {
        if (list.data.length == 0) {
            return "";
        }
        String s = _p(_geti(list.data, 0));
        int i = 1;
        while (i < list.data.length) {
            s = s + " -> " + _p(_geti(list.data, i));
            i = i + 1;
        }
        return s;
    }

    static SinglyLinkedList insert_nth(SinglyLinkedList list, int index, int value) {
        if (index < 0 || index > list.data.length) {
            throw new RuntimeException(String.valueOf("index out of range"));
        }
        int[] res = ((int[])(new int[]{}));
        int i_1 = 0;
        while (i_1 < index) {
            res = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(res), java.util.stream.IntStream.of(list.data[i_1])).toArray()));
            i_1 = i_1 + 1;
        }
        res = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(res), java.util.stream.IntStream.of(value)).toArray()));
        while (i_1 < list.data.length) {
            res = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(res), java.util.stream.IntStream.of(list.data[i_1])).toArray()));
            i_1 = i_1 + 1;
        }
        return new SinglyLinkedList(res);
    }

    static SinglyLinkedList insert_head(SinglyLinkedList list, int value) {
        return insert_nth(list, 0, value);
    }

    static SinglyLinkedList insert_tail(SinglyLinkedList list, int value) {
        return insert_nth(list, list.data.length, value);
    }

    static DeleteResult delete_nth(SinglyLinkedList list, int index) {
        if (index < 0 || index >= list.data.length) {
            throw new RuntimeException(String.valueOf("index out of range"));
        }
        int[] res_1 = ((int[])(new int[]{}));
        int val = 0;
        int i_2 = 0;
        while (i_2 < list.data.length) {
            if (i_2 == index) {
                val = list.data[i_2];
            } else {
                res_1 = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(res_1), java.util.stream.IntStream.of(list.data[i_2])).toArray()));
            }
            i_2 = i_2 + 1;
        }
        return new DeleteResult(new SinglyLinkedList(res_1), val);
    }

    static DeleteResult delete_head(SinglyLinkedList list) {
        return delete_nth(list, 0);
    }

    static DeleteResult delete_tail(SinglyLinkedList list) {
        return delete_nth(list, list.data.length - 1);
    }

    static int get_item(SinglyLinkedList list, int index) {
        if (index < 0 || index >= list.data.length) {
            throw new RuntimeException(String.valueOf("index out of range"));
        }
        return list.data[index];
    }

    static SinglyLinkedList set_item(SinglyLinkedList list, int index, int value) {
        if (index < 0 || index >= list.data.length) {
            throw new RuntimeException(String.valueOf("index out of range"));
        }
        int[] res_2 = ((int[])(new int[]{}));
        int i_3 = 0;
        while (i_3 < list.data.length) {
            if (i_3 == index) {
                res_2 = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(res_2), java.util.stream.IntStream.of(value)).toArray()));
            } else {
                res_2 = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(res_2), java.util.stream.IntStream.of(list.data[i_3])).toArray()));
            }
            i_3 = i_3 + 1;
        }
        return new SinglyLinkedList(res_2);
    }

    static SinglyLinkedList reverse_list(SinglyLinkedList list) {
        int[] res_3 = ((int[])(new int[]{}));
        int i_4 = list.data.length - 1;
        while (i_4 >= 0) {
            res_3 = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(res_3), java.util.stream.IntStream.of(list.data[i_4])).toArray()));
            i_4 = i_4 - 1;
        }
        return new SinglyLinkedList(res_3);
    }

    static void main() {
        SinglyLinkedList lst = empty_list();
        int i_5 = 1;
        while (i_5 <= 5) {
            lst = insert_tail(lst, i_5);
            i_5 = i_5 + 1;
        }
        System.out.println(to_string(lst));
        lst = insert_head(lst, 0);
        System.out.println(to_string(lst));
        DeleteResult del = delete_head(lst);
        lst = del.list;
        System.out.println(_p(del.value));
        del = delete_tail(lst);
        lst = del.list;
        System.out.println(_p(del.value));
        del = delete_nth(lst, 2);
        lst = del.list;
        System.out.println(_p(del.value));
        lst = set_item(lst, 1, 99);
        System.out.println(_p(get_item(lst, 1)));
        lst = reverse_list(lst);
        System.out.println(to_string(lst));
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }

    static Integer _geti(int[] a, int i) {
        return (i >= 0 && i < a.length) ? a[i] : null;
    }
}
